node/libs/crypto/src/bn254/mod.rs

```diff
@@ -0,0 +1,239 @@
+            .map_err(|e| anyhow!("failed to decode secret key: {e:?}"))
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        let mut buf = Vec::new();
+        self.0.serialize_compressed(&mut buf).unwrap();
+        buf
+    }
+}
+
+/// Type safety wrapper around G2.
+#[derive(Clone)]
+pub struct PublicKey(pub G2);
+
+impl PartialEq for PublicKey {
+    fn eq(&self, other: &Self) -> bool {
+        ByteFmt::encode(self).eq(&ByteFmt::encode(other))
+    }
+}
+
+impl Eq for PublicKey {}
+
+impl ByteFmt for PublicKey {
+    fn decode(bytes: &[u8]) -> anyhow::Result<Self> {
+        G2::deserialize_compressed(bytes)
+            .map(Self)
+            .map_err(|e| anyhow!("failed to decode public key: {e:?}"))
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        let mut buf = Vec::new();
+        self.0.serialize_compressed(&mut buf).unwrap();
+        buf
+    }
+}
+
+impl std::hash::Hash for PublicKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.hash(state);
+    }
+}
+
+impl PartialOrd for PublicKey {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for PublicKey {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        ByteFmt::encode(self).cmp(&ByteFmt::encode(other))
+    }
+}
+
+/// Type safety wrapper around a G1 value.
+#[derive(Clone, Debug)]
+pub struct Signature(pub G1);
+
+impl Signature {
+    /// Verifies a signature against the provided public key
+    pub fn verify(&self, msg: &[u8], pk: &PublicKey) -> Result<(), Error> {
+        let hash_point = hash::hash_to_g1(msg);
+
+        // First pair: e(H(m): G1, pk: G2)
+        let a = Bn254::pairing(hash_point, pk.0);
+        // Second pair: e(sig: G1, generator: G2)
+        let b = Bn254::pairing(self.0, G2::generator());
+
+        if a == b {
+            Ok(())
+        } else {
+            Err(Error::SignatureVerificationFailure)
+        }
+    }
+}
+
+impl PartialEq for Signature {
+    fn eq(&self, other: &Self) -> bool {
+        ByteFmt::encode(self).eq(&ByteFmt::encode(other))
+    }
+}
+
+impl Eq for Signature {}
+
+impl ByteFmt for Signature {
+    fn decode(bytes: &[u8]) -> anyhow::Result<Self> {
+        G1::deserialize_compressed(bytes)
+            .map(Self)
+            .map_err(|e| anyhow!("failed to decode signature: {e:?}"))
+    }
+    fn encode(&self) -> Vec<u8> {
+        let mut buf = Vec::new();
+        self.0.serialize_compressed(&mut buf).unwrap();
+        buf
+    }
+}
+
+impl PartialOrd for Signature {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Signature {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        ByteFmt::encode(self).cmp(&ByteFmt::encode(other))
+    }
+}
+/// Type safety wrapper around [Signature] indicating that it is an aggregated signature.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct AggregateSignature(Signature);
```

Comment:
```suggestion
pub struct AggregateSignature(pub G1);
```
It would be preferable to store the point directly.

Reply:
Fixed.